# Agent Customization & Prompt Engineering in AutoGen v0.4

In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('OPENAI_API_KEY')
model_client = OpenAIChatCompletionClient(model='gpt-4', api_key=api_key)

### Agent Customization

- We can assign a role to our agent
- help in fitting agent to specific use case


In [2]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history. Provide detailed and accuragte answers about historical events,figures and timelines'
)

In [3]:
async def test_history_expert():
    result = await asssistant.run(task = 'Who was the first President of USA?')
    print(result.messages[-1].content)
    

In [4]:
await test_history_expert()

The first President of the United States was George Washington. He was born on February 22, 1732, in Westmoreland County, Virginia. Washington held the presidency for two terms from 1789 to 1797. Before becoming President, he played prominent roles in the military, leading the American Continental Army against the British during the American Revolutionary War and was also one of the Founding Fathers of the United States. As a leader, he is often remembered for voluntarily resigning from the presidency after two terms, setting a precedent for future presidents. He passed away on December 14, 1799, at his plantation, Mount Vernon, in Virginia.


The first President of the United States was George Washington. He served from 1789 to 1797. Before his presidency, Washington was well-known for his role in the American Revolutionary War as the commander-in-chief of the Continental Army. He is also noted for his presidency because he set many of the precedents for the office that are still followed today, such as limiting himself to only two terms in office. His presidency is significant for establishing numerous structures and procedures that still exist in the U.S. government today.

# Prompt Engineering

[Prompt Engineering - OpenAI ](https://platform.openai.com/docs/guides/prompt-engineering)

In [5]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history who explaines event in a storytelling style, as if narrating a historic novel. Be vivid and engaging and focus on figures and events'
)

In [6]:
async def test_storytelling():
    result = await asssistant.run(task = 'Tell me about something on the signing of Declaration of Independence')
    print(result.messages[-1].content)
    

In [7]:
await test_storytelling()

Ah, let's spin the wheel of time back to the sweltering summer of 1776. Picture this: the 13 American states are ailing under British rule; taxation without representation, an idea that could curdle milk. But a new dawn is about to break, one ripe with revolution and ready for freedom.

In the heart of Philadelphia, a robust brick building stands tall. This is the Pennsylvania State House, now known as the Independence Hall, a place humming with anticipation and high-ridden nerves. Huddled inside, are 56 men, mostly lawyers, plantation owners, merchants; representatives from each state, carrying the burden of their people's hopes. Laden on their shoulders is the monumental task of drafting a document that would sever ties with the British Crown, and would echo through the centuries. 

Thomas Jefferson, a young delegate from Virginia with an eloquence matched by few, has been tasking himself with penning down the formal justification for independence. A fireplace flickers beside him, ca

In [8]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

# Load API key from .env
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
model_client = OpenAIChatCompletionClient(model="gpt-4o", api_key=api_key)

# Define a simple tool
def get_weather(city: str) -> str:
    return f"The weather in {city} is sunny!"  # Dummy response

# Set up the agent
agent = AssistantAgent(
    name="weather_agent",
    model_client=model_client,
    system_message="You are a weather assistant. Use the get_weather tool when asked about weather.",
    tools=[get_weather],
    reflect_on_tool_use=True
)

# Run the agent
async def main():
    result = await agent.run(task="What’s the weather in New York?")
    print(result.messages[-1].content)

await main()


The weather in New York is sunny!
